### PR TITLE
Dialog: Add Dialog Stickiness

### PR DESCRIPTION
**Pull request by jspurlin**
#### Pull request checklist

- [X] Include a change request file using `$ npm run change`

#### Description of changes
This PR add the ability for dialogs to be sticky/modeless (e.g. the user can interact with the content behind the dialog while the dialog is still open).

This is an example of the new behavior:
modelessdialog

NOTE: this still requires creators to manage the presence of one (or more) dialogs at a time, but will not fail if more than one dialog is present at the same time.

#### Focus areas to test
Verified sticky dialogs work as expected in IE, Edge, Chrome, and FF. Verified that by default if non-sticky dialog is launched while a sticky dialog is also launched that the non-sticky (blocking or not) dialog wins.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8065)

### CODE REVIEW

**Comment by jspurlin** [comment]
FYI @SavannahBourgeois

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts [hunk 1]
```diff
@@ -63,6 +63,13 @@ export interface IModalProps extends React.ClassAttributes<ModalBase>, IWithResp
    */
   isBlocking?: boolean;
 
+  /**
+   * Whether the dialog should be sticky (e.g. not dismiss when focusing/clicking outside of the dialog).
+   * if true: isBlocking is ignored, there will be no overlay (isDarkOverlay is ignored),
+   * isClickableOutsideFocusTrap is true, and forceFocusInsideTrap is false
+   */
+  isSticky?: boolean;
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: I think I would prefer `isModeless` but that also highlights another issue: `Modal` should inherently be modal. I think a modeless modal is just a dialog. Should `Modal` support this prop? Can `Dialog` be used instead?

But now looking at `Modal` and `Dialog` implementations I see that `Dialog` actually uses `Modal`. I would have expected that to be the other way around. A `Modal` is a specialized blocking version of `Dialog`. I would have expected `Modal` to use `Dialog`. 😕 

TL;DR: I guess my conclusion is that I would prefer `isModeless` wording as more standardized and descriptive.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts [hunk 2]
```diff
@@ -111,10 +118,13 @@ export type IModalStyleProps = Required<Pick<IModalProps, 'theme'>> &
     hasBeenOpened?: boolean;
     /** Positioning of modal on first render */
     modalRectangleTop?: number;
+    /** Whether the modal is sticky (can interact with content behind the dialog without dismissal) */
+    isSticky?: boolean;
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: I would move this to the `Pick` on line 112 to avoid redundant definitions that can get out of sync. (looks like a few of the others should be moved too)

File: packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts [hunk 3]
```diff
@@ -55,6 +55,13 @@ export interface ILayerProps extends React.HTMLAttributes<HTMLDivElement | Layer
    * default while allowing users to opt in to the new event bubbling functionality.
    */
   eventBubblingEnabled?: boolean;
+
+  /**
+   * Whether the layer should be added as the first child of the host.
+   * If true, the layer will be inserted as the first child of the host
+   * By default, the layer will be appended at the end to the host
+   */
+  insertAsHostFirstChild?: boolean;
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts]: This naming seems a bit cumbersome. Can we do something simpler, like `insertFirst`? I think there's enough implication given Layer context that it implies where the Layer will be inserted.

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts]: Sure

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 1 above)
Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: Yeah, the same thoughts went through my head (hence why I called it sticky instead of modeless). I'll update it to modeless

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 2 above)
Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: i.e.
```tsx
Pick<IModalProps, 'className' | 'containerClassName' | 'scrollableContentClassName' | 'topOffsetFixed' | 'isSticky'> 
```

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: I updated mine but didn't change any of the other ones, like isOpen, because it is the name of a prop but what's really being passed is the isOpen state

**Comment by jspurlin** [comment]
It's strange that the CI build is failing, I updated the Modal snapshot (and it's happy when I run this locally)

**Comment by KevinTCoughlin** [comment]
> It's strange that the CI build is failing, I updated the Modal snapshot (and it's happy when I run this locally)

It may be worth merging `master` to see if it unblocks CI.

**Comment by jspurlin** [comment]
...PING... @JasonGore and @ThomasMichon

File: packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts [hunk 4]
```diff
@@ -55,6 +55,13 @@ export interface ILayerProps extends React.HTMLAttributes<HTMLDivElement | Layer
    * default while allowing users to opt in to the new event bubbling functionality.
    */
   eventBubblingEnabled?: boolean;
+
+  /**
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts]: Perhaps this should be `insertMode: 'append' | 'prepend'`, where `'append'` is the default.

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts]: I don't have a strong preference, but if we go the `'append' | 'prepend'` approach we have to use defaultProps which I'm never a fan of if it can be avoided (with a bool where undefined is falsy, for example)

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts]: @ThomasMichon , any strong preference here? I'd like to get this merged, but want to make sure you agree :)

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Layer/Layer.types.ts]: @ThomasMichon I'm going to merge this today unless you have anything to add
